corehq/motech/generic_inbound/utils.py

```diff
@@ -63,58 +65,56 @@ def from_request(cls, request):
         except UnicodeDecodeError:
             raise GenericInboundUserError(_("Unable to decode request body"))
 
-        try:
-            request_json = json.loads(body)
-        except json.JSONDecodeError:
-            raise GenericInboundUserError(_("Payload must be valid JSON"))
         return cls(
             domain=request.domain,
             couch_user=request.couch_user,
             request_method=request.method,
             user_agent=request.META.get('HTTP_USER_AGENT'),
-            data=request_json,
+            data=body,
             query=dict(request.GET.lists()),
-            headers=get_headers_for_api_context(request)
+            headers=get_headers_for_api_context(request),
+            request_id=request_id or uuid.uuid4().hex,
         )
 
     @classmethod
     def from_log(cls, log):
-        try:
-            request_json = json.loads(log.request_body)
-        except (UnicodeDecodeError, json.JSONDecodeError):
-            raise GenericInboundUserError(_("Payload must be valid JSON"))
         return cls(
             domain=log.domain,
             couch_user=CouchUser.get_by_username(log.username),
             request_method=log.request_method,
             user_agent=log.request_headers.get('HTTP_USER_AGENT'),
-            data=request_json,
+            data=log.request_body,
             query=dict(QueryDict(log.request_query).lists()),
             headers=dict(log.request_headers),
+            request_id=log.id
         )
 
-    def to_context(self):
-        restore_user = self.couch_user.to_ota_restore_user(
+    @cached_property
+    def restore_user(self):
+        return self.couch_user.to_ota_restore_user(
             self.domain, request_user=self.couch_user)
-        return get_evaluation_context(
-            restore_user,
-            self.request_method,
-            self.query,
-            self.headers,
-            self.data,
-        )
 
 
 @attr.s(kw_only=True, frozen=True, auto_attribs=True)
 class ApiResponse:
+    """Data class for managing response data and producing HTTP responses.
+    Override ``_get_http_response`` to return different HTTP response."""
     status: int
-    data: dict = None
+    internal_response: dict = None
+
+    def get_http_response(self):
+        if self.status == 204:
+            return HttpResponse(status=204)  # no body for 204 (RFC 7230)
+        return self._get_http_response()
+
+    def _get_http_response(self):
+        return JsonResponse(self.internal_response, status=self.status)
 
 
 def make_processing_attempt(response, request_log, is_retry=False):
     from corehq.motech.generic_inbound.models import ProcessingAttempt
 
-    response_data = response.data or {}
+    response_data = response.internal_response or {}
```

Comment:
Do you think it's necessary for us to store the final, transformed response as well?  It _should_ be deterministic from the `internal_response`, but I could see it being useful for debugging, especially if we make changes to that JSON-to-HL7 transformation code.

Reply:
I did want to do that but ran out of time.